src/treetensornetworks/projttns/abstractprojttn.jl

```diff
@@ -131,13 +131,43 @@ end
 function position(
   P::AbstractProjTTN{V}, psi::TTN{V}, pos::Union{Vector{<:V},NamedEdge{V}}
 ) where {V}
-  # shift position
   P = shift_position(P, pos)
-  # remove internal edges (out of place)
+  P = invalidate_environments(P)
+  P = make_environments(P, psi)
+  return P
+end
+
+function position!(
+  P::AbstractProjTTN{V}, psi::TTN{V}, pos::Union{Vector{<:V},NamedEdge{V}}
+) where {V}
+  P = shift_position(P, pos)
+  P = invalidate_environments(P)
+  make_environments!(P, psi, e)
+  return P
+end
+
+function invalidate_environments(P::AbstractProjTTN)
   ie = internal_edges(P)
   newenvskeys = filter(!in(ie), keys(P.environments))
-  P = ProjTTN(pos, P.H, getindices(P.environments, newenvskeys))
-  # make all environments surrounding new position
+  P = ProjTTN(P.pos, P.H, getindices(P.environments, newenvskeys))
```

Comment:
Something we can do to clean this up a bit and make it more generic to other `AbstractProjTTN` types is define a generic function `set_environments`, for example:
```julia
set_environments(p::AbstractProjTTN, environments) = error("Not implemented")

set_environments(p::ProjTTN, environments) = ProjTTN(p.pos, p.H, environments)
```
I like to define those kind of field setter functions for all of my types, it makes it easier to modify immutable objects. I also like using the package [Accessors.jl](https://github.com/JuliaObjects/Accessors.jl) to make it easier to define these kind of field setter functions.

Reply:
I defined setter and getter functions. Since [Accessors.jl](https://github.com/JuliaObjects/Accessors.jl) isn't used in ITensorNetworks I think it's better not to introduce it here?

Reply:
Yeah, that's fine, we can start using it in a future PR. Just pointing it out for future reference since I think it is a good way of thinking about designing types.